### PR TITLE
Replace Cross::CreateDir with create_dir function

### PR DESCRIPTION
**Pull request by dreamer**
```
    New create_dir function behaves exactly like mkdir, so it can be used
    for replacing all mkdir usage as well (which means removal of ifdefs, as
    Unix mkdir is different than Windows mkdir/_mkdir).
    
    This change requires adding some additional checks in code that was
    missing it (where Cross::CreateDir was being used) - this code generated
    warnings in static analyzer tools. This change allows us to provide
    better logs or error messages to users.
```

This is a cleanup required before implementing the last feature for FluidSynth we want to support for 0.76.0 (searching sf2 files in existing directories). It's required so we could provide `soundfonts` dir behaving the same way as `glshaders` behaves right now.

### CODE REVIEW

File: src/gui/sdlmain.cpp
Comment (kcgen) [inline src/gui/sdlmain.cpp:3069]: It would be nice to encapsulate this logic in `create_dir`, to avoid duplicate these post-fail checks "do we actually have the directory or was it blocked because a filename exists in it's place?".

`create_dir` could take a boolean with a meaning like "pass if directory exists".

File: src/gui/sdlmain.cpp
Comment (kcgen) [inline src/gui/sdlmain.cpp:3073]: It would be nice to avoid the `char desc[STRERR_LEN];` whenever we want to use strerror. 

If we're wrapping it, can the description be included (static char[])?

An interface like:
- `std::string safe_strerror(desc)` or
- `const char * safe_strerror(desc)` 

Would be cleaner and less to remember.

File: src/gui/sdlmain.cpp
Comment (dreamer) [reply src/gui/sdlmain.cpp]: Creating `dir_exists` would open the gate for a plethora of additional functions like `file_exists`, `link_exists`, etc - I don't want to do that.

Adding boolean parameters to functions is a [code smell](https://alexkondov.com/should-you-pass-boolean-to-functions/), so I preferred to avoid it; I also wanted to make the function a drop-in replacement.

Decided to add an optional `flags` parameter, with only a single flag supported for now (but I can easily imagine adding few more - e.g. option for recursive creation of directories).

File: src/gui/sdlmain.cpp
Comment (dreamer) [reply src/gui/sdlmain.cpp]: Ehh, I wanted to avoid that so we could transition to C11 `strerror_s` in the future, but it's even more complicated :/

Implemented `std::string safe_strerror(errno)` but I have no idea if it works outside of Linux, we need testing on Windows, macOS and perhaps a BSD...

File: src/gui/sdlmain.cpp
Comment (kcgen) [reply src/gui/sdlmain.cpp]: > Adding boolean parameters to functions is a code smell

I agree; and as usual your implementation is top-notch!  

If and when the need arises, having more flags available (and the ability to combine them elegantly) will produce very readable code.

File: src/gui/sdlmain.cpp
Comment (kcgen) [reply src/gui/sdlmain.cpp]: This also looks great! 

I guess Google's test framework is a small nightmare to get working on macOS and Windows.

 I can do some testing on Windows. Maybe some macOS community members can lend their help.

File: tests/fs_utils.cpp
Comment (kcgen) [inline tests/fs_utils.cpp:77]: Interesting; this is how to create a pre-requisite action before each test?

**Review by kcgen — APPROVED** [review]
Excellent fine-grained refactoring @dreamer!

The code is more readable and another swath of preprocessor branches are eliminated. 🧹

File: src/gui/sdlmain.cpp
Comment (dreamer) [reply src/gui/sdlmain.cpp]: On macOS: I simply haven't tried yet - it seems like it is missing from brew. On Windows it's not hard to get going via VS, but a whole new project needs to be added to solution - I have this work started on an older branch.

Additionally, it seems like gtest maintainers changed their mind and now they recommend bundling source code of test framework with the repository :/ maybe we'll use a Git module for this in the future, if I'll be able to make it work with VS.

Testing is simple: 
```
rm -rf ~/.config/dosbox/capture
touch ~/.config/dosbox/capture
./src/dosbox -opencaptures echo
```
Correct output should say something like this on stderr: `Can't access capture dir '/home/dreamer_/.config/dosbox/capture': File exists`

File: tests/fs_utils.cpp
Comment (dreamer) [reply tests/fs_utils.cpp:77]: Yup; it's called a "test fixture" (notice `TEST_F` macro is used below). Actions pre/post-test can be placed in c-tor/d-tor (modern approach) or in SetUp/TearDown methods (legacy approach). Doing it this way guarantees that cleanup happens even if test fails, so it won't cause cascading test failures due to leaving stuff behind.

**Comment by dreamer** [comment]
Thanks for the review @kcgen - having another set of eyes looking at this definitely helped :)

Later I'll test with Windows and if it passes, I think we'll be ok to merge. I am somewhat worried about GNU-specific version of `strerror_r` - it works ok on Linux, but I have no idea what and where injects `_GNU_SOURCE` into our build.

**Comment by dreamer** [comment]
Tested on Windows - works as expected.

**Pull request merged**
